homeassistant/components/amberelectric/coordinator.py

```diff
@@ -0,0 +1,149 @@
+"""Amber Electric Coordinator."""
+from __future__ import annotations
+
+from datetime import timedelta
+
+from amberelectric import ApiException
+from amberelectric.api import amber_api
+from amberelectric.model.actual_interval import ActualInterval
+from amberelectric.model.channel import ChannelType
+from amberelectric.model.current_interval import CurrentInterval
+from amberelectric.model.forecast_interval import ForecastInterval
+from amberelectric.model.site import Site
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import LOGGER
+
+
+def is_current(interval: ActualInterval | CurrentInterval | ForecastInterval) -> bool:
+    """Return true if the supplied interval is a CurrentInterval."""
+    return interval.__class__ == CurrentInterval
+
+
+def is_forecast(interval: ActualInterval | CurrentInterval | ForecastInterval) -> bool:
+    """Return true if the supplied interval is a ForecastInterval."""
+    return interval.__class__ == ForecastInterval
+
+
+def is_general(interval: ActualInterval | CurrentInterval | ForecastInterval) -> bool:
+    """Return true if the supplied interval is on the general channel."""
+    return interval.channel_type == ChannelType.GENERAL
+
+
+def is_controlled_load(
+    interval: ActualInterval | CurrentInterval | ForecastInterval,
+) -> bool:
+    """Return true if the supplied interval is on the controlled load channel."""
+    return interval.channel_type == ChannelType.CONTROLLED_LOAD
+
+
+def is_feed_in(interval: ActualInterval | CurrentInterval | ForecastInterval) -> bool:
+    """Return true if the supplied interval is on the feed in channel."""
+    return interval.channel_type == ChannelType.FEED_IN
+
+
+def first(
+    intervals: ActualInterval | CurrentInterval | ForecastInterval,
+) -> ActualInterval | CurrentInterval | ForecastInterval | None:
+    """Return the first element if it exists, otherwise return None."""
+    if len(intervals) > 0:
+        return intervals[0]
+    return None
+
+
+class AmberDataService:
```

Comment:
Have this class extend DataUpdateCoordinator. 

Instantiate this coordinator in your `__init__.py` file and use `coordiantor.async_config_entry_first_refresh()` to automatically trigger ConfigEntryNotReady etc.